### PR TITLE
Improve and update the C# basics page for 3.1 beta 3

### PR DESCRIPTION
**Pull request by Calinou**
This simplifies the Mono installation instructions, as it is now only required for NuGet and MSBuild (as of 3.1 beta 3, Godot bundles its own Mono 5.18 installation).

I'd like to have seasoned C# users review this, to see if all this makes sense :slightly_smiling_face: The last section in the end in particular sounds confusing to me, but I rephrased it the best I could.

### CODE REVIEW

File: getting_started/scripting/c_sharp/c_sharp_basics.rst [hunk 1]
```diff
@@ -3,93 +3,122 @@
...
-.. note:: To download Mono v5.12+ on a Mac, locate the "Stable Channel" link from the `Mono Downloads Page <http://www.mono-project.com/download/>`_. The Visual Studio channel is an earlier version of Mono and will not work with Godot 3.0.3+.
+Setting up C# for Godot
+-----------------------
 
-You also need MSBuild (at least version 15.0), which should come with the Mono installation.
+Download and install the latest version of the
+`Mono SDK <http://www.mono-project.com/download/>`_. As of Godot 3.1 beta 3,
+the version number doesn't matter since Godot bundles its own Mono 5.18
+installation. We only need the Mono installation for NuGet and MSBuild
+which are required to use C# in Godot.
```
Comment (neikeq) [inline getting_started/scripting/c_sharp/c_sharp_basics.rst]: Mono's MSBuild gives issues on Windows. We recently changed the default build tool on Windows to Visual Studio's MSBuild. The user needs to either install Visual Studio 2017 or `Build Tools for Visual Studio 2017 ` if they don't want the entire IDE (https://visualstudio.microsoft.com/downloads/).
The plan is to bundle MSBuild as well but that will take some time.

File: getting_started/scripting/c_sharp/c_sharp_basics.rst [hunk 2]
```diff
@@ -3,93 +3,122 @@
...
+so make sure to download the **Mono version** of Godot.
+If you are building Godot from source, make sure to follow the steps to
+enable Mono support in your build as outlined in the
+:ref:`doc_compiling_with_mono` page.
 
-If you are building Godot from source, make sure to follow the steps to include Mono support in your build outlined on the :ref:`doc_compiling_with_mono` page.
-
-In summary, you must have installed 1) the correct Mono SDK version for your Godot version, and 2) The Mono version of Godot.
+In summary, you must have installed the correct Mono SDK version for your
+Godot version **and** the Mono-enabled version of Godot.
```
Comment (neikeq) [inline getting_started/scripting/c_sharp/c_sharp_basics.rst]: This contradicts what is said previously:

>As of Godot 3.1 beta 3,
the version number doesn't matter since Godot bundles its own Mono 5.18
installation.

File: getting_started/scripting/c_sharp/c_sharp_basics.rst [hunk 3]
```diff
@@ -3,93 +3,122 @@
...
+or MonoDevelop. These provide auto-completion, debugging and other features
+useful when working with C#. To set it up in Godot, click on
+**Editor → Editor Settings** and scroll down to the bottom to the
+**Mono** settings. Under **Mono**, click on **Editor** then choose
+your external editor of choice.
 
-.. note:: If you are using Visual Studio Code, ensure you download and install the .NET tools extension. Without this, Godot will crash when trying to create a new C# project or edit a C# script.
+.. note:: If you are using Visual Studio Code, ensure you download and install
+          the .NET Tools extension. Without this, Godot will crash when trying
+          to create a new C# project or edit a C# script.
```
Comment (neikeq) [inline getting_started/scripting/c_sharp/c_sharp_basics.rst]: I don't know of any .NET Tools extension for VSCode. The recommended C# extension if this: https://marketplace.visualstudio.com/items?itemName=ms-vscode.csharp Also Godot shoudn't crash because of this.

File: getting_started/scripting/c_sharp/c_sharp_basics.rst [hunk 4]
```diff
@@ -3,93 +3,122 @@
...
-
-Note that currently, there are some issues where Godot and the C# project don't stay in sync; if you delete, rename or move things, like scripts or nodes, they may no longer match up.
-In this case, it can help to edit the solution files manually.
-
-Example: If you created a script (e.g. ``Test.cs``) and delete it in Godot, the compilation will fail because now the missing file is still expected to be there by the CS project.
-For now, you can simply open up the ``.csproj`` file and look for the ``ItemGroup``, there should be a line included like the following:
+When you create the first C# script, Godot initializes the C# project files
+for your Godot project. This includes generating a C# solution (``.sln``)
+and a project file (``.csproj``), as well as some utility files and folders
+(``.mono``, sometimes ``Properties``).
```
Comment (neikeq) [inline getting_started/scripting/c_sharp/c_sharp_basics.rst]: The `Properties` folder is always generated. It would be better to say something like "and 'Properties/AssemblyInfo.cs' which is common in C# projects". No need to enter in details since people already know what it is if they've used C# in the past.

File: getting_started/scripting/c_sharp/c_sharp_basics.rst [hunk 5]
```diff
@@ -3,93 +3,122 @@
...
+All of these but ``.mono`` are important and should be committed to your
+version control system. ``.mono`` can be safely added to the ignore list of your VCS.
+When troubleshooting, it can sometimes help to delete the ``.mono`` folder
+and let it regenerate.
+
+Note that currently, there are some issues where Godot and the C# project
+don't stay in sync; if you delete, rename or move things like scripts or nodes,
+they may no longer match up. In this case, it can help to edit
+the solution files manually.
```
Comment (neikeq) [inline getting_started/scripting/c_sharp/c_sharp_basics.rst]: The "project file" instead of "solution files".

File: getting_started/scripting/c_sharp/c_sharp_basics.rst [hunk 6]
```diff
@@ -3,93 +3,122 @@
...
+for your Godot project. This includes generating a C# solution (``.sln``)
+and a project file (``.csproj``), as well as some utility files and folders
+(``.mono``, sometimes ``Properties``).
+All of these but ``.mono`` are important and should be committed to your
+version control system. ``.mono`` can be safely added to the ignore list of your VCS.
+When troubleshooting, it can sometimes help to delete the ``.mono`` folder
+and let it regenerate.
+
+Note that currently, there are some issues where Godot and the C# project
+don't stay in sync; if you delete, rename or move things like scripts or nodes,
```
Comment (neikeq) [inline getting_started/scripting/c_sharp/c_sharp_basics.rst]: Moving a node shouldn't cause such an issue.

File: getting_started/scripting/c_sharp/c_sharp_basics.rst (hunk 5 above)
Comment (neikeq) [inline getting_started/scripting/c_sharp/c_sharp_basics.rst]: I will say this in a different way: "if you delete, rename or move a C# script, the change may not be reflected in the C# project file. In cases like this, it will require editing it manually."

File: getting_started/scripting/c_sharp/c_sharp_basics.rst [hunk 7]
```diff
@@ -121,12 +150,16 @@ Here's a blank C# script with some comments to demonstrate how it works.
         }
     }
 
-As you can see, the things normally in global scope in GDScript like Godot's ``print`` function are available in the ``GD`` namespace.
-For a list of those, see the class reference pages for :ref:`@GDScript <class_@gdscript>` and :ref:`@GlobalScope <class_@globalscope>`.
+As you can see, functions normally in global scope in GDScript like Godot's
+``print`` function are available in the ``GD`` namespace.
```
Comment (neikeq) [inline getting_started/scripting/c_sharp/c_sharp_basics.rst]: `GD` is a class that is part of the `Godot` namespace.

It's not needed to enter in details here since the differences are already listed in its own page (to which I have to make some fixes), but I think this should also mention the `Mathf` class.

File: getting_started/scripting/c_sharp/c_sharp_basics.rst [hunk 8]
```diff
@@ -135,33 +168,45 @@ The C# API uses ``PascalCase`` instead of ``snake_case`` in GDScript/C++.
...
+As C# support is quite new in Godot, there are some growing pains and things
+that need to be ironed out. Below is a list of the most important issues
+you should be aware of when diving into C# in Godot, but if in doubt, also
+take a look over the official
+`issue tracker for Mono issues <https://github.com/godotengine/godot/labels/topic%3Amono>`_.
+
+- As explained above, the C# project isn't always kept in sync automatically
+  when things are deleted, renamed or moved in Godot
+  (`#12917 <https://github.com/godotengine/godot/issues/12917>`_).
+- Writing editor plugins and tool scripts in C# is not yet supported.
```
Comment (neikeq) [inline getting_started/scripting/c_sharp/c_sharp_basics.rst]: This is outdated. Tool scripts are supported and so should be editor plugins. The workflow for editor plugins is pretty bad right now, which is worth mentioning, but they should work.
Also, related to this, another point for this list could be: "Currently state is not saved and restored when hot-reloading, with the exception of exported variables."

File: getting_started/scripting/c_sharp/c_sharp_basics.rst [hunk 9]
```diff
@@ -135,33 +168,45 @@ The C# API uses ``PascalCase`` instead of ``snake_case`` in GDScript/C++.
...
+- Exporting Mono projects is only supported for desktop platforms
+  (Linux, Windows and macOS). HTML5, Android, iOS and UWP are not currently supported
+  (`#18364 comment <https://github.com/godotengine/godot/issues/18364#issuecomment-406222102>`_).
```
Comment (neikeq) [inline getting_started/scripting/c_sharp/c_sharp_basics.rst]: The issue link seems unrelated.

File: getting_started/scripting/c_sharp/c_sharp_basics.rst [hunk 10]
```diff
@@ -135,33 +168,45 @@ The C# API uses ``PascalCase`` instead of ``snake_case`` in GDScript/C++.
...
+- Attached C# scripts should refer to a class that has a class name
+  that matches the file name and is unique within the project assembly
+  (`#7402 comment <https://github.com/godotengine/godot/issues/7402#issuecomment-269910926>`_).
```
Comment (neikeq) [inline getting_started/scripting/c_sharp/c_sharp_basics.rst]: The linked issue doesn't make much sense here and my lead to confusion.
About the "unique" thing, I guess this was related to Godot not supporting namespaces back then, which is no longer the case so I would remove that.

**Comment by akien-mga** [comment]
Thanks!
